src/lstm_layer.cpp

```diff
@@ -1494,7 +1682,124 @@ void LSTM::backward(BaseDeltaStates &input_delta_states,
+        this->lstm_states.mu_c_posts.push_back(this->lstm_states.mu_c_prior);
+        this->lstm_states.var_c_posts.push_back(this->lstm_states.var_c_prior);
+        this->lstm_states.mu_zo_priors.push_back(
+            temp_states.linear_states.mu_prior);
+        this->lstm_states.var_zo_priors.push_back(
+            temp_states.linear_states.var_prior);
+        this->lstm_states.mu_zo_posts.push_back(
+            temp_states.linear_states.mu_post);
+        this->lstm_states.var_zo_posts.push_back(
+            temp_states.linear_states.var_post);
+    }
+}
+
+void LSTM::smoother(std::string next_layer_type, BaseTempStates &temp_states)
+/*
+ */
+{
+    int num_timestep = lstm_states.cov_cc.size();
+    int num_hidden_states = lstm_states.cov_cc[0].size();
+
+    lstm_states.mu_c_smooths.resize(num_timestep);
+    lstm_states.var_c_smooths.resize(num_timestep);
+    lstm_states.mu_h_smooths.resize(num_timestep);
+    lstm_states.var_h_smooths.resize(num_timestep);
+    lstm_states.mu_zo_smooths.resize(num_timestep);
+    lstm_states.var_zo_smooths.resize(num_timestep);
+
+    for (auto &val : lstm_states.mu_c_smooths) {
+        val.resize(num_hidden_states, 0.0f);
     }
+    for (auto &val : lstm_states.var_c_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+    for (auto &val : lstm_states.mu_h_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+    for (auto &val : lstm_states.var_h_smooths) {
+        val.resize(num_hidden_states, 0.0f);
+    }
+    for (auto &val : lstm_states.mu_zo_smooths) {
+        val = 0.0f;
+    }
+    for (auto &val : lstm_states.var_zo_smooths) {
+        val = 0.0f;
+    }
+
+    // Initialize the last time step for smoothing
+    lstm_states.mu_c_smooths.back() = lstm_states.mu_c_posts.back();
+    lstm_states.var_c_smooths.back() = lstm_states.var_c_posts.back();
+    lstm_states.mu_h_smooths.back() = lstm_states.mu_h_posts.back();
+    lstm_states.var_h_smooths.back() = lstm_states.var_h_posts.back();
+    lstm_states.mu_zo_smooths.back() = lstm_states.mu_zo_posts.back();
+    lstm_states.var_zo_smooths.back() = lstm_states.var_zo_posts.back();
+
+    lstm_smoother_cell_states(num_timestep, lstm_states.cov_cc,
+                              lstm_states.mu_c_priors, lstm_states.var_c_priors,
+                              lstm_states.mu_c_posts, lstm_states.var_c_posts,
+                              lstm_states.mu_c_smooths,
+                              lstm_states.var_c_smooths);
+
+    lstm_smoother_hidden_states(
+        num_timestep, lstm_states.cov_hc, lstm_states.mu_c_priors,
+        lstm_states.var_c_priors, lstm_states.mu_h_posts,
+        lstm_states.var_h_posts, lstm_states.mu_c_smooths,
+        lstm_states.var_c_smooths, lstm_states.mu_h_smooths,
+        lstm_states.var_h_smooths);
+
+    // only smooth the z_output for the last linear layer
+    if (next_layer_type == "Linear") {
```

Comment:
I would replace this by `LayerType`. See an example [here](https://github.com/lhnguyen102/cuTAGI/blob/main/src/sequential.cpp#L318)